sdk/dotnet/Pulumi/Stack.cs

```diff
@@ -44,30 +43,53 @@ internal sealed class Stack : ComponentResource
         /// <summary>
         /// The outputs of this stack, if the <c>init</c> callback exited normally.
         /// </summary>
-        public readonly Output<IDictionary<string, object?>> Outputs =
+        internal Output<IDictionary<string, object?>> Outputs =
             Output.Create<IDictionary<string, object?>>(ImmutableDictionary<string, object?>.Empty);
 
-        internal Stack(Func<Task<IDictionary<string, object?>>> init)
+        /// <summary>
+        /// Create a Stack with stack resources defined in derived class constructor.
+        /// </summary>
+        public Stack() : this(null)
+        {
+        }
+
+        protected virtual void Initialize() { }
+        protected virtual Task InitializeAsync() => Task.FromResult(0);
```

Comment:
1. Task.CompletedTask
2. doc both of these method.

Reply:
important thing to mention: derived types do not need to call 'base.Initialize'.